core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,215 @@
+                       @Nullable Backoff backoff) {
+        requireNonNull(loader, "loader");
+        this.loader = loader;
+        this.expireAfterLoad = expireAfterLoad;
+        this.expireIf = expireIf;
+        this.exceptionHandler = exceptionHandler;
+        this.backoff = backoff;
+    }
+
+    @Override
+    public CompletableFuture<T> get() {
+        return get0().thenApply(f -> f.loadVal);
+    }
+
+    private CompletableFuture<CacheEntry<T>> get0() {
+        CompletableFuture<CacheEntry<T>> future;
+        CacheEntry<T> cacheEntry = null;
+        for (;;) {
+            final CompletableFuture<CacheEntry<T>> loadFuture = this.loadFuture;
+            if (!loadFuture.isDone()) {
+                return loadFuture;
+            }
+
+            if (!loadFuture.isCompletedExceptionally()) {
+                cacheEntry = loadFuture.join();
+                if (isValid(cacheEntry)) {
+                    return loadFuture;
+                }
+            }
+
+            future = new CompletableFuture<>();
+            if (loadFutureUpdater.compareAndSet(this, loadFuture, future)) {
+                break;
+            }
+        }
+
+        final T cache = cacheEntry != null ? cacheEntry.loadVal : null;
+        final CompletableFuture<CacheEntry<T>> newLoadfuture = future;
+        load(cache, newLoadfuture);
+        return newLoadfuture;
+    }
+
+    private void load(@Nullable T cache, CompletableFuture<CacheEntry<T>> future) {
+        load0(cache, future, 0);
+    }
+
+    private void load0(@Nullable T cache, CompletableFuture<CacheEntry<T>> future, int loadAttemptsSoFar) {
+        try {
+            requireNonNull(loader.apply(cache), "loader.apply() returned null")
+                    .handle((val, cause) -> {
+                        if (cause != null) {
+                            logger.warn("Failed to load a new value from loader: {}. the previous value: {}",
+                                        loader, cache, cause);
+                            handleException(cause, cache, future, loadAttemptsSoFar + 1);
+                        } else {
+                            future.complete(new CacheEntry<>(val));
+                        }
+                        return null;
+                    });
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from loader.apply()", e);
+            handleException(e, cache, future, loadAttemptsSoFar + 1);
+        }
+    }
+
+    private void handleException(Throwable cause, @Nullable T cache,
+                                 CompletableFuture<CacheEntry<T>> future, int loadAttemptsSoFar) {
+        if (backoff != null) {
+            retryWithBackoff(cause, cache, future, loadAttemptsSoFar);
+            return;
+        }
+
+        if (exceptionHandler != null) {
+            handleByExceptionHandler(cause, cache, future);
+            return;
+        }
+
```

Comment:
I guess my previous idea which adding a backoff wasn't that good. The backoff makes the implementation complex. I don't know whether it is better to apply `backoff` after or before `exceptionHandler`.

Hence, what do you think of removing `backoff` in this PR? We may revisit it when there is a request for that.